node/adapter/ipam.py

```diff
@@ -0,0 +1,115 @@
+from netaddr import IPAddress
+from datastore import DatastoreClient
+
+IP_ASSIGNMENT_PATH = "/calico/ipam/%(version)s/assignment/%(pool)s"
+IP_ASSIGNMENT_KEY = IP_ASSIGNMENT_PATH + "/%(address)s"
+
+
+class SequentialAssignment(object):
+    """
+    Assign IP addresses sequentially
+    """
+
+    def __init__(self):
+        # Init an etcd client.
+        self.etcd = IPAMClient()
+
+    def allocate(self, pool):
+        """
+        Attempt to allocate an IP address from the provided pool.
+
+        :param IPNetwork pool: The pool to allocate from
+        :return: An IP address which has been allocated or None
+        if allocation failed.
+        :rtype str:
+        """
+        while True:
+            assigned_addresses = self.etcd.get_assigned_addresses(pool)
+
+            candidate_address = self._get_next(pool, assigned_addresses)
+            if candidate_address is None:
+                # the pool is full, we can't allocate an address
+                return None
+            else:
+                # We've found an address to try.
+                if self.etcd.assign_address(pool,
+                                            IPAddress(candidate_address)):
+                    return candidate_address
+
+    def _get_next(self, pool, assigned):
+        """
+        Gets the next address in a range.
+        :param IPNetwork pool: The pool to allocate from
+        :return: the next IP address to try (a string), or None if the pool
+                 is full.
+        """
+        for addr in pool.iter_hosts():
+            addr_string = str(addr)
+            if addr_string not in assigned:
+                return addr_string
+        return None
+
+
+class IPAMClient(DatastoreClient):
+    def assign_address(self, pool, address):
+        """
+        Attempt to assign an IPAddress in a pool.
+        Fails if the address is already assigned.
+        The directory for storing assignments in this pool must already exist.
+
+        :param IPNetwork pool: The pool that the assignment is from.
+        :param IPAddress address: The address to assign.
+
+        :return: True if the allocation succeeds, false otherwise.
```

Comment:
This isn't very Pythony.   You should throw an exception instead.  This is useful because there are lots of different problems we might signal in the future:
 - etcd down
 - pool didn't exist
 - address wasn't available,
etc.